Add `system_info(atom_limit)`

Gsantomaggio opened the pull request:
Add `system_info(atoms_limit)` to provide a way to retrieve the
maximum number of atoms allowed. Add tests and documentation for it too.

```
➜ ✗ bin/erl +t 1123456
Erlang/OTP 19 [erts-8.1] [source-e895301] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V8.1  (abort with ^G)
1> erlang:system_info(atoms_limit).
1123456
2>
```
Added:
 - [atoms_list](https://github.com/Gsantomaggio/otp/blob/system_info_atoms/erts/preloaded/src/erlang.erl#L2511) on `erlang.erl` 
 - [test](https://github.com/Gsantomaggio/otp/blob/system_info_atoms/erts/emulator/test/system_info_SUITE.erl#L532)
 
test result:
 ```
 1> ts:run(emulator, system_info_SUITE, atoms_limit, [batch]).
...
Common Test starting (cwd is /Users/gabriele/github/Personal/otp_atom_limit/release/tests/emulator_test)
Common Test: Running make in test directories...

CWD set to: "/Users/gabriele/github/Personal/otp_atom_limit/release/tests/test_server/ct_run.test_server@mac.2016-10-10_14.27.50"

TEST INFO: 1 test(s), 1 case(s) in 1 suite(s)

Testing tests.emulator_test.system_info_SUITE.atoms_limit: Starting test, 1 test cases
Testing tests.emulator_test.system_info_SUITE.atoms_limit: TEST COMPLETE, 1 ok, 0 failed of 1 test cases

Updating /Users/gabriele/github/Personal/otp_atom_limit/release/tests/test_server/index.html ... done
Updating /Users/gabriele/github/Personal/otp_atom_limit/release/tests/test_server/all_runs.html ... done
ok
2>
```

Also refactored [start_node](https://github.com/Gsantomaggio/otp/blob/system_info_atoms/erts/emulator/test/system_info_SUITE.erl#L516) to avoid code duplication

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

garazdawi commented:
the functionality and tests works, the docs need some more work.

1) The section of atom_limit should be in the same section as ets_limit, i.e. the large one and not in the allocator section where it is now. 
2) As you've partially noticed the clauses in a function are referenced by their absolute id. That means that when you insert a new one at slot 6 you have to increment all the clause_i fields after it. You've done it in a couple of places but not all. For instance the ets_limit spec now shows up in the allocator section, there are a couple of more examples of this. The best/only way to check that it is correct is to build the html docs and manually check that all clauses seems to be where they should be.

Also please squash all your commits into a single commit when you are done.

Gsantomaggio commented:
Sorry for the documentation errors.

I updated the document, now it is in the same `ets_limit` section:

![header](https://cloud.githubusercontent.com/assets/386987/19316059/540f71e4-90a0-11e6-8df6-413cfcfdd451.png)



and here is the body, with the `+t ` link :


![description](https://cloud.githubusercontent.com/assets/386987/19316086/6944adb8-90a0-11e6-81a7-113aa8331a6d.png)


<details>
 <summary>Here is the diff html output  file</summary>

```
--- /Users/gabriele/github/erlang_original_doc/erts-8.1/doc/html/erlang_original.html	Wed Oct 12 12:16:17 2016
+++ /Users/gabriele/github/erlang_original_doc/erts-8.1/doc/html/erlang_ATOM.html	Wed Oct 12 16:56:12 2016
@@ -8475,6 +8475,7 @@
               information returned by
               <span class="bold_code"><a href="#system_info_allocator_tuple"><span class="code">erlang:system_info({allocator,Alloc})</span></a></span>.</p>
           </dd>
+          
         </dl>
       </p></div>
 
@@ -8583,7 +8584,7 @@
         </dl>
       </p></div>
 
-    <p><span class="bold_code">erlang:system_info(Item :: fullsweep_after) -&gt;<br>                      {fullsweep_after, integer() &gt;= 0}</span><br><span class="bold_code">erlang:system_info(Item :: garbage_collection) -&gt;<br>                      [{atom(), integer()}]</span><br><span class="bold_code">erlang:system_info(Item :: max_heap_size) -&gt;<br>                      {max_heap_size,<br>                       MaxHeapSize :: max_heap_size()}</span><br><span class="bold_code">erlang:system_info(Item :: message_queue_data) -&gt;<br>                      <span class="bold_code"><a href="#type-message_queue_data">message_queue_data()</a></span></span><br><span class="bold_code">erlang:system_info(Item :: min_heap_size) -&gt;<br>                      {min_heap_size,<br>                       MinHeapSize :: integer() &gt;= 1}</span><br><span class="bold_code">erlang:system_info(Item :: min_bin_vheap_size) -&gt;<br>                      {min_bin_vheap_size,<br>                       MinBinVHeapSize :: integer() &gt;= 1}</span><br><div class="REFBODY">
+    <p><span class="bold_code">erlang:system_info(Item :: fullsweep_after) -&gt;<br>                      {fullsweep_after, integer() &gt;= 0}</span><br><span class="bold_code">erlang:system_info(Item :: machine) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: max_heap_size) -&gt;<br>                      {max_heap_size,<br>                       MaxHeapSize :: max_heap_size()}</span><br><span class="bold_code">erlang:system_info(Item :: message_queue_data) -&gt;<br>                      <span class="bold_code"><a href="#type-message_queue_data">message_queue_data()</a></span></span><br><span class="bold_code">erlang:system_info(Item :: min_heap_size) -&gt;<br>                      {min_heap_size,<br>                       MinHeapSize :: integer() &gt;= 1}</span><br><span class="bold_code">erlang:system_info(Item :: min_bin_vheap_size) -&gt;<br>                      {min_bin_vheap_size,<br>                       MinBinVHeapSize :: integer() &gt;= 1}</span><br><span class="bold_code">erlang:system_info(Item :: modified_timing_level) -&gt;<br>                      integer() | undefined</span><br><div class="REFBODY">
 <p>Types:</p>
 <div class="REFTYPES">
       <span class="bold_code">message_queue_data() = off_heap | on_heap</span><br>
@@ -8655,7 +8656,7 @@
         </dl>
       </p></div>
 
-    <p><span class="bold_code">erlang:system_info(Item :: build_type) -&gt;<br>                      opt |<br>                      debug |<br>                      purify |<br>                      quantify |<br>                      purecov |<br>                      gcov |<br>                      valgrind |<br>                      gprof |<br>                      lcnt |<br>                      frmptr</span><br><span class="bold_code">erlang:system_info(Item :: c_compiler_used) -&gt; {atom(), term()}</span><br><span class="bold_code">erlang:system_info(Item :: check_io) -&gt; [term()]</span><br><span class="bold_code">erlang:system_info(Item :: compat_rel) -&gt; integer()</span><br><span class="bold_code">erlang:system_info(Item :: creation) -&gt; integer()</span><br><span class="bold_code">erlang:system_info(Item :: debug_compiled) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: delayed_node_table_gc) -&gt;<br>                      infinity | integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dirty_cpu_schedulers) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dirty_cpu_schedulers_online) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dirty_io_schedulers) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dist) -&gt; binary()</span><br><span class="bold_code">erlang:system_info(Item :: dist_buf_busy_limit) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dist_ctrl) -&gt;<br>                      {Node :: node(),<br>                       ControllingEntity :: port() | pid()}</span><br><span class="bold_code">erlang:system_info(Item :: driver_version) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: dynamic_trace) -&gt;<br>                      none | dtrace | systemtap</span><br><span class="bold_code">erlang:system_info(Item :: dynamic_trace_probes) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: elib_malloc) -&gt; false</span><br><span class="bold_code">erlang:system_info(Item :: eager_check_io) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: ets_limit) -&gt; integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: heap_sizes) -&gt; [integer() &gt;= 0]</span><br><span class="bold_code">erlang:system_info(Item :: heap_type) -&gt; private</span><br><span class="bold_code">erlang:system_info(Item :: info) -&gt; binary()</span><br><span class="bold_code">erlang:system_info(Item :: kernel_poll) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: loaded) -&gt; binary()</span><br><span class="bold_code">erlang:system_info(Item ::<br>                       logical_processors |<br>                       logical_processors_available |<br>                       logical_processors_online) -&gt;<br>                      unknown | integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: machine) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: modified_timing_level) -&gt;<br>                      integer() | undefined</span><br><span class="bold_code">erlang:system_info(Item :: multi_scheduling) -&gt;<br>                      disabled |<br>                      blocked |<br>                      blocked_normal |<br>                      enabled</span><br><span class="bold_code">erlang:system_info(Item :: multi_scheduling_blockers) -&gt;<br>                      [Pid :: pid()]</span><br><span class="bold_code">erlang:system_info(Item :: nif_version) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: normal_multi_scheduling_blockers) -&gt;<br>                      [Pid :: pid()]</span><br><span class="bold_code">erlang:system_info(Item :: otp_release) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: os_monotonic_time_source) -&gt;<br>                      [{atom(), term()}]</span><br><span class="bold_code">erlang:system_info(Item :: os_system_time_source) -&gt;<br>                      [{atom(), term()}]</span><br><span class="bold_code">erlang:system_info(Item :: port_count) -&gt; integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: port_limit) -&gt; integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: process_count) -&gt; integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: process_limit) -&gt; integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: procs) -&gt; binary()</span><br><span class="bold_code">erlang:system_info(Item :: scheduler_bind_type) -&gt;<br>                      spread |<br>                      processor_spread |<br>                      thread_spread |<br>                      thread_no_node_processor_spread |<br>                      no_node_processor_spread |<br>                      no_node_thread_spread |<br>                      no_spread |<br>                      unbound</span><br><span class="bold_code">erlang:system_info(Item :: scheduler_bindings) -&gt; tuple()</span><br><span class="bold_code">erlang:system_info(Item :: scheduler_id) -&gt;<br>                      SchedulerId :: integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: schedulers | schedulers_online) -&gt;<br>                      integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: smp_support) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: start_time) -&gt; integer()</span><br><span class="bold_code">erlang:system_info(Item :: system_version) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: system_architecture) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: threads) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: thread_pool_size) -&gt; integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: time_correction) -&gt; true | false</span><br><span class="bold_code">erlang:system_info(Item :: time_offset) -&gt;<br>                      preliminary | final | volatile</span><br><span class="bold_code">erlang:system_info(Item :: time_warp_mode) -&gt;<br>                      no_time_warp |<br>                      single_time_warp |<br>                      multi_time_warp</span><br><span class="bold_code">erlang:system_info(Item :: tolerant_timeofday) -&gt;<br>                      enabled | disabled</span><br><span class="bold_code">erlang:system_info(Item :: trace_control_word) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: update_cpu_info) -&gt; changed | unchanged</span><br><span class="bold_code">erlang:system_info(Item :: version) -&gt; string()</span><br></p>
+    <p><span class="bold_code">erlang:system_info(Item :: atoms_limit) -&gt; integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: build_type) -&gt;<br>                      opt |<br>                      debug |<br>                      purify |<br>                      quantify |<br>                      purecov |<br>                      gcov |<br>                      valgrind |<br>                      gprof |<br>                      lcnt |<br>                      frmptr</span><br><span class="bold_code">erlang:system_info(Item :: c_compiler_used) -&gt; {atom(), term()}</span><br><span class="bold_code">erlang:system_info(Item :: check_io) -&gt; [term()]</span><br><span class="bold_code">erlang:system_info(Item :: compat_rel) -&gt; integer()</span><br><span class="bold_code">erlang:system_info(Item :: creation) -&gt; integer()</span><br><span class="bold_code">erlang:system_info(Item :: debug_compiled) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: delayed_node_table_gc) -&gt;<br>                      infinity | integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dirty_cpu_schedulers) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dirty_cpu_schedulers_online) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dirty_io_schedulers) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dist) -&gt; binary()</span><br><span class="bold_code">erlang:system_info(Item :: dist_buf_busy_limit) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: dist_ctrl) -&gt;<br>                      {Node :: node(),<br>                       ControllingEntity :: port() | pid()}</span><br><span class="bold_code">erlang:system_info(Item :: driver_version) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: dynamic_trace) -&gt;<br>                      none | dtrace | systemtap</span><br><span class="bold_code">erlang:system_info(Item :: dynamic_trace_probes) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: elib_malloc) -&gt; false</span><br><span class="bold_code">erlang:system_info(Item :: eager_check_io) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: ets_limit) -&gt; integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: garbage_collection) -&gt;<br>                      [{atom(), integer()}]</span><br><span class="bold_code">erlang:system_info(Item :: heap_sizes) -&gt; [integer() &gt;= 0]</span><br><span class="bold_code">erlang:system_info(Item :: heap_type) -&gt; private</span><br><span class="bold_code">erlang:system_info(Item :: info) -&gt; binary()</span><br><span class="bold_code">erlang:system_info(Item :: kernel_poll) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: loaded) -&gt; binary()</span><br><span class="bold_code">erlang:system_info(Item ::<br>                       logical_processors |<br>                       logical_processors_available |<br>                       logical_processors_online) -&gt;<br>                      unknown | integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: min_bin_vheap_size) -&gt;<br>                      {min_bin_vheap_size,<br>                       MinBinVHeapSize :: integer() &gt;= 1}</span><br><span class="bold_code">erlang:system_info(Item :: multi_scheduling) -&gt;<br>                      disabled |<br>                      blocked |<br>                      blocked_normal |<br>                      enabled</span><br><span class="bold_code">erlang:system_info(Item :: multi_scheduling_blockers) -&gt;<br>                      [Pid :: pid()]</span><br><span class="bold_code">erlang:system_info(Item :: nif_version) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: normal_multi_scheduling_blockers) -&gt;<br>                      [Pid :: pid()]</span><br><span class="bold_code">erlang:system_info(Item :: otp_release) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: os_monotonic_time_source) -&gt;<br>                      [{atom(), term()}]</span><br><span class="bold_code">erlang:system_info(Item :: os_system_time_source) -&gt;<br>                      [{atom(), term()}]</span><br><span class="bold_code">erlang:system_info(Item :: port_count) -&gt; integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: port_limit) -&gt; integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: process_count) -&gt; integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: process_limit) -&gt; integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: procs) -&gt; binary()</span><br><span class="bold_code">erlang:system_info(Item :: scheduler_bind_type) -&gt;<br>                      spread |<br>                      processor_spread |<br>                      thread_spread |<br>                      thread_no_node_processor_spread |<br>                      no_node_processor_spread |<br>                      no_node_thread_spread |<br>                      no_spread |<br>                      unbound</span><br><span class="bold_code">erlang:system_info(Item :: scheduler_bindings) -&gt; tuple()</span><br><span class="bold_code">erlang:system_info(Item :: scheduler_id) -&gt;<br>                      SchedulerId :: integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: schedulers | schedulers_online) -&gt;<br>                      integer() &gt;= 1</span><br><span class="bold_code">erlang:system_info(Item :: smp_support) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: start_time) -&gt; integer()</span><br><span class="bold_code">erlang:system_info(Item :: system_version) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: system_architecture) -&gt; string()</span><br><span class="bold_code">erlang:system_info(Item :: threads) -&gt; boolean()</span><br><span class="bold_code">erlang:system_info(Item :: thread_pool_size) -&gt; integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: time_correction) -&gt; true | false</span><br><span class="bold_code">erlang:system_info(Item :: time_offset) -&gt;<br>                      preliminary | final | volatile</span><br><span class="bold_code">erlang:system_info(Item :: time_warp_mode) -&gt;<br>                      no_time_warp |<br>                      single_time_warp |<br>                      multi_time_warp</span><br><span class="bold_code">erlang:system_info(Item :: tolerant_timeofday) -&gt;<br>                      enabled | disabled</span><br><span class="bold_code">erlang:system_info(Item :: trace_control_word) -&gt;<br>                      integer() &gt;= 0</span><br><span class="bold_code">erlang:system_info(Item :: update_cpu_info) -&gt; changed | unchanged</span><br><span class="bold_code">erlang:system_info(Item :: version) -&gt; string()</span><br></p>
 <div class="REFBODY"><p>
         <p>Returns various information about the current system
           (emulator) as specified by <span class="code">Item</span>:</p>
@@ -8664,6 +8665,16 @@
             <span class="code">alloc_util_allocators</span>, <span class="code">allocator_sizes</span></strong></dt>
           <dd>
             <p>See <span class="bold_code"><a href="#system_info_allocator_tags">above</a></span>.</p>
+          </dd>
+          <dt><strong><span class="code">atoms_limit</span></strong></dt>
+          <dd>
+            <a name="system_info_atoms_limit"></a>
+            <p>Returns the maximum number of atoms allowed.
+              This limit can be increased at startup by passing
+              command-line flag
+              <span class="bold_code"><a href="javascript:erlhref('../../../doc/../','erts','erl.html#+t');"><span class="code">+t</span></a></span> to
+              <span class="code">erl(1)</span>. 
+            </p>
           </dd>
           <dt><strong><span class="code">build_type</span></strong></dt>
           <dd>
```
</details>

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

Gsantomaggio commented:
@garazdawi the branch is ready.

psyeugenic commented:
Are we sure about the name? `atoms_limit` vs. `atom_limit` vs. anything something else?

Gsantomaggio commented:
maybe `atom_limit` is better, as the other ones (not plurals).

IngelaAndin commented:
atom_limit sounds more natural to me.

Gsantomaggio commented:
done with `atom_limit`:

```
➜ /bin/erl +t 12345690
Erlang/OTP 19 [erts-8.1] [source-a5b13a3] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V8.1  (abort with ^G)
1> erlang:system_info(atom_limit).
12345690
2>
```

OTP-Maintainer commented:
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: add erlang:system_info(atom_limit).


----------------

I am a script, I am not human 

----------------

Gsantomaggio commented:
Commit message fixed.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

garazdawi commented:
merged into master
